### ISSUE TITLE
Single.takeUntil does not dispose subscription to 'other' upon disposal

### ISSUE THREAD

**Issue by damianw** [ReactiveX/RxJava#6018]
Using `Single.takeUntil` does not dispose the subscription it makes to `other` when the main subscription is disposed. RxJava Version: `2.1.13`

#### Sample
```Kotlin
val control = Single.never<Nothing>()
        .doOnSubscribe { println("Subscribe control") }
        .doOnDispose { println("Dispose control") }
val disposable = Single.never<Nothing>()
        .doOnSubscribe { println("Subscribe main") }
        .doOnDispose { println("Dispose main") }
        .takeUntil(control)
        .subscribe()
disposable.dispose()
```
prints only:
```
Subscribe control
Subscribe main
Dispose main
```

#### Comparison with Observable
For comparison, `Observable.takeUntil`:
```Kotlin
val control = Observable.never<Nothing>()
        .doOnSubscribe { println("Subscribe control") }
        .doOnDispose { println("Dispose control") }
val disposable = Observable.never<Nothing>()
        .doOnSubscribe { println("Subscribe main") }
        .doOnDispose { println("Dispose main") }
        .takeUntil(control)
        .subscribe()
disposable.dispose()
```

prints:
```
Subscribe control
Subscribe main
Dispose main
Dispose control
```

**Comment by akarnokd** [comment]
Thanks for reporting. This is indeed a bug with `Single.takeUntil(Single)` and with a bunch of other variants. I'll post a fix shortly.

**Comment by akarnokd** [comment]
Fix posed in #6019. Since the whole `Single.takeUntil` is bugged, the workaround is to convert both to `Observable` for the duration of the call:

```kotlin
val control = ...

val main = ...

main.toObservable().takeUntil(control.toObservable()).singleOrError()
```
